src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -368,89 +370,117 @@ ErrorIfQueryNotSupported(Query *queryTree)
+					 "filter on joining columns.";
+	char *filterHint = "Consider using equal filter on partition column.";
 
 	if (queryTree->hasSubLinks)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Subqueries other than in from-clause are currently unsupported";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->hasWindowFuncs)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Window functions are currently unsupported";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->setOperations)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Union, Intersect, or Except are currently unsupported";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->hasRecursive)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Recursive queries are currently unsupported";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->cteList)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Common Table Expressions are currently unsupported";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->hasForUpdate)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "For Update/Share commands are currently unsupported";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->distinctClause)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Distinct clause is currently unsupported";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->groupingSets)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Grouping sets, cube, and rollup is currently unsupported";
+		errorHint = filterHint;
 	}
 
 	hasTablesample = HasTablesample(queryTree);
 	if (hasTablesample)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Tablesample is currently unsupported";
+		errorHint = filterHint;
 	}
 
 	hasUnsupportedJoin = HasUnsupportedJoinWalker((Node *) queryTree->jointree, NULL);
 	if (hasUnsupportedJoin)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Join types other than inner/outer joins are currently unsupported";
+		errorHint = joinHint;
 	}
 
 	hasComplexJoinOrder = HasComplexJoinOrder(queryTree);
 	if (hasComplexJoinOrder)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Complex join orders are currently unsupported";
+		errorHint = joinHint;
 	}
 
 	hasComplexRangeTableType = HasComplexRangeTableType(queryTree);
 	if (hasComplexRangeTableType)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Complex table expressions are currently unsupported";
+		errorHint = filterHint;
 	}
 
+
 	/* finally check and error out if not satisfied */
 	if (!preconditionsSatisfied)
 	{
-		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-						errmsg("cannot perform distributed planning on this query"),
-						errdetail("%s", errorDetail)));
+		bool showHint = ErrorHintRequired(errorHint, queryTree);
+		if (showHint)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot perform distributed planning on this query"),
+							errdetail("%s", errorDetail),
+							errhint("%s", errorHint)));
```

Comment:
This is one of the few places I'd rather see a ternary expression. For these messages, PostgreSQL has the idiom `showHint ? errhint("%s", errorHint), 0` (see `ReportRemoteError` for our usage of this idiom).